shiny-ci-cd.Rmd

```diff
@@ -34,9 +34,72 @@ Generally speaking, a CI/CD pipeline related to an R package is comprised of the
+[Workflows](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions) use `YAML` syntax and should be stored in the `.github/workflows` directory in the root of the repository.
+Workflows are constituted of jobs and each job is a set of steps to perform individual tasks, e.g. commands or actions.
+
+### Workflow steps
+
+A workflow should have an identifying `name` and an `on` section that indicates upon which events the workflow should be triggered. It should include at least one job and each job will have a set of steps fully specifying what to execute. Such steps can be an action (predefined, sourcing from GitHub repos that contain such actions) or a script. However, for the time being, GitHub Actions does not support a hierarchical, aggregated structure of actions. Being entirely customizable, it is necessary to fully specify each step in the CI/CD pipeline explicitly.
+
+#### Setup
+
+- Checkout the source package from the current repo, using `actions/checkout` provided by GitHub.
+- Setup R using the action [`r-lib/actions/setup-r`](https://github.com/r-lib/actions#readme).
+- Query and cache R package dependencies using `remotes::dev_package_deps()` and the `actions/cache` predefined action.
+- Install system dependencies using package `sysreqs::sysreq_commands()` (for the `ubuntu` runner used to run the action).
+- Install R package dependencies using `remotes::Install_deps()`.
+
+#### Package check
+
+- Check the package via `rcmdcheck::rcmdcheck()`.
+
+#### Deployment
+
+- Deploy to shinyapps.io:
+  - In this case, environment variables `SHINYAPPS_ACCOUNT`, `SHINYAPPS_TOKEN` and `SHINYAPPS_SECRET` (defining credentials for [shinyapps.io](https://www.shinyapps.io/)), are specified / accessible as GitHub [secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference#authentication-and-secrets).
+  - An R script e.g. `deploy/deploy-shinyapps.R` (build-ignored via `usethis::use_build_ignore("deploy")`) defines the deployment commands based on the environment variables:
+```{r read-deploy-script, eval = TRUE, echo = FALSE}
+knitr::read_chunk("shiny-ci-cd/deploy/deploy-shinyapps.R", labels = "deploy-shinyapps")
+```
+```{r deploy-shinyapps}
+```
+
+### Workflow file
+
+The `steps` described in the previous section are defined in the `.yml` workflow file as follows:
+
+```{r read-ci-cd.yml, eval = TRUE, include = FALSE}
+file.copy("shiny-ci-cd", "_book", recursive = TRUE)
+knitr::read_chunk("shiny-ci-cd/actions/ci-cd.yml", labels = "ci-cd.yml")
+```
+```{yml ci-cd.yml}
+```
+
+As visible from the run logs that can be found on GitHub under the `Actions` tab, all the CI/CD pipeline steps are performed subsequently, and are identifiable by the `name` field. See the example below:
+
+![GitHub Actions Continuous Integration / Continuous Deployment pipeline for a packaged Shiny app](shiny-ci-cd/img/ShinyCICD_githubactions1.png)
+
```

Comment:
We completely miss the part about `renv`, which is very relevant to us. We can introduce this new section in the very same way as currently done in the Travis part (perhaps making that part slimmer do avoid redundancy), and highlight which steps are different (caching and restoring locked dependencies). Not sure we want to show the full worfklow, maybe just the `ci-renv.yml`, since deployment is unaffected

Reply:
This requires some thinking, as including the full ci-cd-renv.yml file just after ci-cd.yml seems a bit excessive (same with including a new screenshot for the renv steps). We could use collapsible sections for this or even put it there entirely, not sure what looks more appealing.

Reply:
I suggest to go ahead with the introduction, just mentionthat `renv` requires a different caching and that dependencies are installed via `renv::restore`, pointing to https://rstudio.github.io/renv/articles/ci.html#github-actions.

I would not include new screenshots here, and would avoid the the full workflow in the `renv` section. Still, I think it makes sense to perhaps enhance the section about `usethis` to give more relevance to the 4 full workflow .yml files we provide and show them, indeed using collapsible / accordion. We could do this in R markdown using bsplus (http://ijlyttle.github.io/bsplus/index.html), see how tit was done for the CovidR contest in  https://github.com/Milano-R/erum2020-covidr-contest/blob/a13b0ee2b2eec57bc523c7d6e50734863e35a9d8/contest/index.Rmd#L86-L103

Overall we could have stg like 

```md

### Complete workflows and `usethis::use_github_action()`

Full YAML workflows for CI and CI/CD pipelines, with and without `renv` are shown below and provided as part of this guide.

In order to setup and use CI/CD GitHub Actions workflows as described above, you can simply include the relevant workflow file your project via:

[...] existing code snippet

[...] collapsible panel for each workflow files (perhaps within a dedicated `####` sub section), the panel (and subsection) title being the file name and the content is the file content

Reply:
@riccardoporreca I updated part of this, but could not add collapsible sections using bsplus (nor other approaches I tried from some quick googling + trying out) - it seems that collapsible sections don't work with bookdown and / or gitbook.

Before I do a more thorough investigation, would you know something about this? Maybe some work-around with html and possibly other things? I was searching for something about "collapsible sections" in the bookdown documentation, but could not find anything helpful regarding this problem.

Reply:
- Collapsible full workflow files included using the `<details>` (https://www.w3schools.com/tags/tag_details.asp) in commit 19b455d 
- Explicit section about renv under GitHub action included as hinted above in commit 900444e